File: kotlin-node/karakum/src/jsMain/resources/plugins/convertSymlinkType.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isIdentifier(node)
        && node.text === "Type"

        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)

        && node.parent.parent
        && ts.isModuleBlock(node.parent.parent)

        && node.parent.parent.parent
        && ts.isModuleDeclaration(node.parent.parent.parent)
        && ts.isIdentifier(node.parent.parent.parent.name)
        && node.parent.parent.parent.name.text === "symlink"
    ) {
        return "SymlinkType"
    }

    if (
        ts.isQualifiedName(node)
        && ts.isIdentifier(node.left)
        && node.left.text === "symlink"
        && node.right.text === "Type"
    ) {
        return "SymlinkType"
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertSymlinkType.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.isIdentifier
import typescript.isModuleBlock
import typescript.isModuleDeclaration
import typescript.isQualifiedName
import typescript.isTypeAliasDeclaration

val convertSymlinkType = createPlugin { node, _, _ ->
    nullable {
        ensure(isIdentifier(node))

        val typeAlias = ensureNotNull(node.getParentOrNull())
        ensure(isTypeAliasDeclaration(typeAlias))

        val moduleBlock = ensureNotNull(typeAlias.getParentOrNull())
        ensure(isModuleBlock(moduleBlock))

        val module = ensureNotNull(moduleBlock.getParentOrNull())
        ensure(isModuleDeclaration(module))

        val moduleName = module.name
        ensure(isIdentifier(moduleName))
        ensure(moduleName.text == "symlink")

        "SymlinkType"
    } ?: nullable {
        ensure(isQualifiedName(node))

        val left = node.left
        ensure(isIdentifier(left))
        ensure(left.text == "symlink")

        ensure(node.right.text == "Type")

        "SymlinkType"
    }
}
